src/app-layer-frames.c

```diff
@@ -0,0 +1,416 @@
+                if (x >= FRAMES_STATIC_CNT) {
+                    nframe = &frames->dframes[x - FRAMES_STATIC_CNT];
+                } else {
+                    nframe = &frames->sframes[x];
+                }
+                StreamDPUCopy(nframe, frame);
+                nframe->rel_offset -= slide; /* turns negative if start if before window */
+                if (frame != nframe) {
+                    FrameClean(frame);
+                }
+                x++;
+            }
+        }
+    }
+    frames->cnt = x;
+    AppLayerFrameDumpForFrames("post_slide", frames);
+    return 0;
+}
+
+void AppLayerFramesUpdateProgress(
+        Flow *f, TcpStream *stream, const uint64_t progress, const uint8_t diframetion)
+{
+    FramesContainer *frames_container = AppLayerFramesGetContainer(f);
+    if (frames_container == NULL)
+        return;
+
+    Frames *frames;
+    if (diframetion == STREAM_TOSERVER) {
+        frames = &frames_container->toserver;
+    } else {
+        frames = &frames_container->toclient;
+    }
+
+    const uint32_t slide = progress - STREAM_APP_PROGRESS(stream);
+    frames->progress_rel += slide;
+}
+
+void AppLayerFramesSlide(Flow *f, const uint32_t slide, const uint8_t diframetion)
+{
+    FramesContainer *frames_container = AppLayerFramesGetContainer(f);
+    if (frames_container == NULL)
+        return;
+    Frames *frames;
+    if (diframetion == STREAM_TOSERVER) {
+        frames = &frames_container->toserver;
+    } else {
+        frames = &frames_container->toclient;
+    }
+    FrameSlide(frames, slide);
+}
+
+static void FrameFreeSingleFrame(Frames *frames, Frame *r)
+{
+    FrameDebug("free", frames, r);
+    FrameClean(r);
+}
+
+void FramesFree(Frames *frames)
+{
+    BUG_ON(frames == NULL);
+
+    for (uint16_t i = 0; i < frames->cnt; i++) {
+        if (i < FRAMES_STATIC_CNT) {
+            Frame *r = &frames->sframes[i];
+            FrameFreeSingleFrame(frames, r);
+        } else {
+            const uint16_t o = i - FRAMES_STATIC_CNT;
+            Frame *r = &frames->dframes[o];
+            FrameFreeSingleFrame(frames, r);
+        }
+    }
+    SCFree(frames->dframes);
+    frames->dframes = NULL;
+}
+
+Frame *AppLayerFrameNew(Flow *f, const StreamSlice *stream_slice, const uint8_t *frame_start,
+        const uint32_t len, int dir, uint8_t frame_type)
+{
+    SCLogDebug("stream_slice offset %" PRIu64, stream_slice->offset);
+    SCLogDebug("frame_start %p stream_slice->input %p", frame_start, stream_slice->input);
+
+    /* workarounds for many (unit|fuzz)tests not handling TCP data properly */
+#if defined(UNITTESTS) || defined(FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION)
+    if (f->protoctx == NULL)
+        return NULL;
+    if (frame_start < stream_slice->input || frame_start >= stream_slice->input + stream_slice->input_len)
+        return NULL;
+#endif
+    BUG_ON(frame_start < stream_slice->input);
+    BUG_ON(stream_slice->input == NULL);
+    BUG_ON(f->proto != IPPROTO_TCP);
```

Comment:
No UDP support? I got tripped up while looking at DNS frame support which can be TCP or UDP.

Reply:
> Looking at the logging for a crude DNS conversion I did, I see this logged:
> ...
> Out side of debug, is the `id` needed? Same with `offset`, and to a lesser extend `length`.

Offset I think tells you something useful: its the start of the frame vs the start of the stream. Id I'm not sure about whether its useful.

> Also, interesting field, `diframetion`? Would the direction be needed with that other work that Eric has been doing that tells us more about the direction of an alert?

That should be `direction`. Its a artifact of `s/rec/frame/g` ;)

Reply:
I've focused on TCP for now, as this is where the perf gain is to be had, so blocked everything else. But I don't think its hard to extend this to UDP. The data the frames point to will be in the actual packet payload then.